asciidtype/asciidtype/src/casts.c

```diff
@@ -72,33 +76,243 @@ ascii_to_ascii(PyArrayMethod_Context *context, char *const data[],
+                                     PyArray_DTypeMeta *NPY_UNUSED(dtypes[2]),
+                                     PyArray_Descr *given_descrs[2],
+                                     PyArray_Descr *loop_descrs[2],
+                                     npy_intp *NPY_UNUSED(view_offset))
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+    long in_size = ((ASCIIDTypeObject *)given_descrs[0])->size;
+    if (given_descrs[1] == NULL) {
+        PyArray_Descr *unicode_descr = PyArray_DescrNewFromType(NPY_UNICODE);
+        // numpy stores unicode as UCS4 (4 bytes wide), so bitshift
+        // by 2 to get the number of bytes needed to store the UCS4 charaters
+        unicode_descr->elsize = in_size << 2;
+        loop_descrs[1] = unicode_descr;
+    }
+    else {
+        Py_INCREF(given_descrs[1]);
+        loop_descrs[1] = given_descrs[1];
+    }
+
+    long out_size = (loop_descrs[1]->elsize) >> 2;
+
+    if (out_size >= in_size) {
+        return NPY_SAFE_CASTING;
+    }
+
+    return NPY_UNSAFE_CASTING;
+}
+
 static PyArray_DTypeMeta *a2a_dtypes[2] = {NULL, NULL};
 
 static PyType_Slot a2a_slots[] = {
         {NPY_METH_resolve_descriptors, &ascii_to_ascii_resolve_descriptors},
-        {_NPY_METH_get_loop, &ascii_to_ascii_get_loop},
+        {NPY_METH_strided_loop, &ascii_to_ascii},
+        {NPY_METH_unaligned_strided_loop, &ascii_to_ascii},
         {0, NULL}};
 
 PyArrayMethod_Spec ASCIIToASCIICastSpec = {
         .name = "cast_ASCIIDType_to_ASCIIDType",
         .nin = 1,
         .nout = 1,
-        .flags = NPY_METH_SUPPORTS_UNALIGNED,
-        .casting = NPY_SAME_KIND_CASTING,
+        .casting = NPY_UNSAFE_CASTING,
+        .flags = (NPY_METH_NO_FLOATINGPOINT_ERRORS |
+                  NPY_METH_SUPPORTS_UNALIGNED),
         .dtypes = a2a_dtypes,
         .slots = a2a_slots,
 };
+
+static PyType_Slot u2a_slots[] = {
+        {NPY_METH_resolve_descriptors, &unicode_to_ascii_resolve_descriptors},
+        {NPY_METH_strided_loop, &unicode_to_ascii},
+        {0, NULL}};
+
+static char *u2a_name = "cast_Unicode_to_ASCIIDType";
+
+static PyType_Slot a2u_slots[] = {
+        {NPY_METH_resolve_descriptors, &ascii_to_unicode_resolve_descriptors},
+        {NPY_METH_strided_loop, &ascii_to_unicode},
+        {0, NULL}};
+
+static char *a2u_name = "cast_ASCIIDType_to_Unicode";
+
+PyArrayMethod_Spec **
+get_casts(void)
+{
+    PyArray_DTypeMeta **u2a_dtypes = malloc(2 * sizeof(PyArray_DTypeMeta *));
+    u2a_dtypes[0] = &PyArray_UnicodeDType;
+    u2a_dtypes[1] = NULL;
+
+    PyArrayMethod_Spec *UnicodeToASCIICastSpec =
+            malloc(sizeof(PyArrayMethod_Spec));
+
+    UnicodeToASCIICastSpec->name = u2a_name;
+    UnicodeToASCIICastSpec->nin = 1;
+    UnicodeToASCIICastSpec->nout = 1;
+    UnicodeToASCIICastSpec->casting = NPY_UNSAFE_CASTING,
+    UnicodeToASCIICastSpec->flags =
+            (NPY_METH_NO_FLOATINGPOINT_ERRORS | NPY_METH_REQUIRES_PYAPI);
```

Comment:
I didn’t know about `NPY_METH_REQUIRES_PYAPI` until I happened to come across it in the experiemental dtype header today. Out of curiosity, what are the downsides of not releasing the GIL in a casting or ufunc loop? For this one I only need a python API function for error handling, so I could re-acquire the GIL manually in the error condition as I originally had it.

Reply:
You want to release the GIL as much as possible (except for very small amount of work, since there is a cost to releasing; although maybe that cost got reduced also).

Reply:
N.B.: we are relying (right now) on the fact that in CPython (and pypy cpyext) accessing the object is OK even without the GIL so long we only get the `elsize`, etc.
For general Python implementations, this may need to happen in the setup, or we have to think about how the `descrs` are passed exactly.